// restructured/lib/Elements.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _Type = require('./Type');

var _Type2 = _interopRequireDefault(_Type);

var _ParserUtil = require('./ParserUtil');

var _ParserUtil2 = _interopRequireDefault(_ParserUtil);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Elements = {};

function capitalize(str) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

_lodash2.default.forEach(_Type2.default.valueTypes, function (type, name) {
  Elements[capitalize(name)] = function () {
    function _class(_ref) {
      var value = _ref.value;
      var position = _ref.position;

      _classCallCheck(this, _class);

      this.type = type;
      this.value = value;
      this.position = position;
    }

    return _class;
  }();
});

_lodash2.default.forEach(_Type2.default.parentTypes, function (type, name) {
  Elements[capitalize(name)] = function () {
    function _class2(_ref2) {
      var _ref2$children = _ref2.children;
      var children = _ref2$children === undefined ? [] : _ref2$children;
      var position = _ref2.position;
      var _ref2$blanklines = _ref2.blanklines;
      var blanklines = _ref2$blanklines === undefined ? [] : _ref2$blanklines;
      var _ref2$bullet = _ref2.bullet;
      var bullet = _ref2$bullet === undefined ? null : _ref2$bullet;
      var _ref2$depth = _ref2.depth;
      var depth = _ref2$depth === undefined ? null : _ref2$depth;
      var _ref2$role = _ref2.role;
      var role = _ref2$role === undefined ? null : _ref2$role;
      var _ref2$indent = _ref2.indent;
      var indent = _ref2$indent === undefined ? null : _ref2$indent;
      var _ref2$directive = _ref2.directive;
      var directive = _ref2$directive === undefined ? null : _ref2$directive;

      _classCallCheck(this, _class2);

      this.type = type;

      if (type === 'bullet_list') {
        this.bullet = bullet;
      }
      if (type === 'interpreted_text') {
        this.role = role;
      }
      if (type === 'section') {
        this.depth = depth;
      }
      if (type === 'directive') {
        this.directive = directive;
      }

      this.position = position;
      if (indent) {
        this.indent = indent;
      }
      this.blanklines = blanklines || [];
      this.children = children;
    }

    return _class2;
  }();
});

Elements.EnumeratorSequence = function () {
  function _class3(_ref3) {
    var type = _ref3.type;
    var value = _ref3.value;

    _classCallCheck(this, _class3);

    this.type = type;
    this.value = value;
  }

  _createClass(_class3, [{
    key: 'width',
    value: function width() {
      return this.value.length;
    }
  }, {
    key: 'isAuto',
    value: function isAuto() {
      return this.type === 'auto';
    }
  }, {
    key: 'isArabicNumerals',
    value: function isArabicNumerals() {
      return this.type === 'arabic_numerals';
    }
  }, {
    key: 'isUppercaseAlphabet',
    value: function isUppercaseAlphabet() {
      return this.type === 'uppercase_alphabet';
    }
  }, {
    key: 'isLowercaseAlphabet',
    value: function isLowercaseAlphabet() {
      return this.type === 'lowercase_alphabet';
    }
  }, {
    key: 'isUppercaseRoman',
    value: function isUppercaseRoman() {
      return this.type === 'uppercase_roman';
    }
  }, {
    key: 'isLowercaseRoman',
    value: function isLowercaseRoman() {
      return this.type === 'lowercase_roman';
    }
  }]);

  return _class3;
}();

Elements.Enumerator = function () {
  function _class4(_ref4) {
    var sequence = _ref4.sequence;
    var format = _ref4.format;

    _classCallCheck(this, _class4);

    this.sequence = sequence;
    this.format = format;
  }

  _createClass(_class4, [{
    key: 'width',
    value: function width() {
      if (this.format === 'parentheses') {
        return this.sequence.width() + 2;
      }
      return this.sequence.width() + 1;
    }
  }, {
    key: 'isPeriodFormat',
    value: function isPeriodFormat() {
      return this.format === 'period';
    }
  }, {
    key: 'isParenthesesFormat',
    value: function isParenthesesFormat() {
      return this.format === 'parentheses';
    }
  }, {
    key: 'isRightParenthesisFormat',
    value: function isRightParenthesisFormat() {
      return this.format === 'right_parenthesis';
    }
  }, {
    key: 'isNext',
    value: function isNext(e) {
      if (this.format !== e.format) {
        return false;
      } else if (e.sequence.isAuto()) {
        return true;
      } else if (this.sequence.type !== e.sequence.type) {
        return false;
      } else if (this.sequence.isArabicNumerals()) {
        return parseInt(this.sequence.value, 10) + 1 === parseInt(e.sequence.value, 10);
      } else if (this.sequence.isUppercaseAlphabet() || this.sequence.isLowercaseAlphabet()) {
        return this.sequence.value.charCodeAt(0) + 1 === e.sequence.value.charCodeAt(0);
      } else if (this.sequence.isUppercaseRoman() || this.sequence.isLowercaseRoman()) {
        return _ParserUtil2.default.romanToNumber(this.sequence.value.toUpperCase()) + 1 === _ParserUtil2.default.romanToNumber(e.sequence.value.toUpperCase());
      }
      return true; // auto
    }
  }, {
    key: 'isFirst',
    value: function isFirst() {
      if (this.sequence.isUppercaseRoman() || this.sequence.isLowercaseRoman()) {
        var value = this.sequence.value;
        return value === 'i' || value === 'I' || value.length === 2;
      }
      return true;
    }
  }]);

  return _class4;
}();

exports.default = Elements;
